A1_perception:
  node:
    period_ms: 10
    climb_condition:
      pitch_alpha: 0.001
      enable_climbing_threshold: -1.15 
      disable_climbing_threshold: -1.15
      timeout_ms: 1800
    inputs:
      # "필수" [From nav2]
      robot_position:           
        topic: "/amcl_pose"
        type: "PoseWithCovarianceStamped"

      # "필수" [From everybot(mcu)]
      imu:
        topic: "/imu_data"
        type: "sensor_msgs/msg/Imu"

      # "필수" [From everybot(lidar)]
      lidar_pc_baselink: 
        topic: "/scan_merged_pointcloud"
        type: "PointCloud2"

      #[From everybot(sensor_manager)]
      collision: 
        topic: "/sensor_to_pointcloud/collision"
        type: "PointCloud2"
        
      #[From everybot(sensor_manager)]
      camera_points: 
        topic: "/sensor_to_pointcloud/camera_object"
        type: "PointCloud2"

    #[From everybot(sensor_manager)]
    multi_tof_inputs:
      left: [
        1,3,5,7,
        17,19,21,23,
        49,51,53,55,
        57,59,61,63
      ]
      right: [
        0,2,4,6,
        16,18,20,22,
        48,50,52,54,
        56,58,60,62
      ]
      
    layers:
      # one_d_tof:
      #   inputs:
      #   - "one_d_tof"
      #   filter:
      #     compose:
      #       filters:
      #         timeout:
      #           timeout_milliseconds: 60000
      #         lidar_dist_check:
      #           min_range: 0.0
      #           max_range: 0.1
      #         one_d_roi:
      #           use_inside: true
      #           polygon: [ [0.41, -0.5], [0.41, 0.5], [0.18,0.5], [0.18,-0.50]]
      #         one_d_tof:
      #           use_stop: true
      #           stop_and_back_dist: 0.33
      #           line_length: 0.10
      #           resolution: 0.05

      # camera_points:
      #   inputs:
      #   - "camera_points"
      #   filter:
      #     compose:
      #       filters:
      #         timeout:
      #           timeout_milliseconds: 60000
      #         sector_roi:
      #           use_inside: true
      #           min_range: 0.3
      #           max_range: 5.0
      #           min_angle: -180
      #           max_angle: 180
      #         density:
      #           max_count: 3
      #           radius: 0.2            

      # collision:
      #   inputs:
      #   - "collision"
      #   filter:
      #     compose:
      #       filters:
      #         collision:
      #           line_length: 0.1
      #           resolution: 0.05


######## persistence & real time

      # multi_tof_13rows:
      #   inputs:
      #   - "multi_tof_left_idx_1"
      #   - "multi_tof_left_idx_3"
      #   - "multi_tof_left_idx_5"
      #   - "multi_tof_left_idx_7"
      #   - "multi_tof_left_idx_17"
      #   - "multi_tof_left_idx_19"
      #   - "multi_tof_left_idx_21"
      #   - "multi_tof_left_idx_23"
      #   - "multi_tof_right_idx_0"
      #   - "multi_tof_right_idx_2"
      #   - "multi_tof_right_idx_4"
      #   - "multi_tof_right_idx_6"
      #   - "multi_tof_right_idx_16"
      #   - "multi_tof_right_idx_18"
      #   - "multi_tof_right_idx_20"
      #   - "multi_tof_right_idx_22"
      #   filter:
      #     compose:
      #       filters:
      #         timeout:
      #           timeout_milliseconds: 15000
      #         lidar_dist_check:
      #           min_range: 0.0
      #           max_range: 0.15
      #         sector_roi:
      #           use_inside: true
      #           min_range: 1.01
      #           max_range: 2.5
      #           min_angle: -160
      #           max_angle: 160
      #         density:
      #           max_count: 3
      #           radius: 0.3

      # multi_tof_13rows_realtime:
      #   inputs:
      #   - "multi_tof_left_idx_1"
      #   - "multi_tof_left_idx_3"
      #   - "multi_tof_left_idx_5"
      #   - "multi_tof_left_idx_7"
      #   - "multi_tof_left_idx_17"
      #   - "multi_tof_left_idx_19"
      #   - "multi_tof_left_idx_21"
      #   - "multi_tof_left_idx_23"
      #   - "multi_tof_right_idx_0"
      #   - "multi_tof_right_idx_2"
      #   - "multi_tof_right_idx_4"
      #   - "multi_tof_right_idx_6"
      #   - "multi_tof_right_idx_16"
      #   - "multi_tof_right_idx_18"
      #   - "multi_tof_right_idx_20"
      #   - "multi_tof_right_idx_22"
      #   filter:
      #     compose:
      #       filters:
      #         timeout:
      #           timeout_milliseconds: 100
      #         lidar_dist_check:
      #           min_range: 0.0
      #           max_range: 0.15
      #         sector_roi:
      #           use_inside: true
      #           min_range: 0.3
      #           max_range: 1.0
      #           min_angle: -50.3
      #           max_angle: 50.3
      #         density:
      #           max_count: 3
      #           radius: 0.1

######## low level & drop off

      multi_tof_left_49_57:
        inputs:
        - "multi_tof_left_idx_49"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.10
              low_obstacle:
                input: "multi_tof_left_idx_57"
                dist_max: 0.5
                dist_min: 0.0

      multi_tof_left_51_59:
        inputs:
        - "multi_tof_left_idx_51"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.10
              low_obstacle:
                input: "multi_tof_left_idx_59"
                dist_max: 0.5
                dist_min: 0.0

      multi_tof_left_53_61:
        inputs:
        - "multi_tof_left_idx_53"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.10
              low_obstacle:
                input: "multi_tof_left_idx_61"
                dist_max: 0.5
                dist_min: 0.0

      multi_tof_left_55_63:
        inputs:
        - "multi_tof_left_idx_55"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.10
              low_obstacle:
                input: "multi_tof_left_idx_63"
                dist_max: 0.5
                dist_min: 0.0

      multi_tof_right_48_56:
        inputs:
        - "multi_tof_right_idx_48"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.10
              low_obstacle:
                input: "multi_tof_right_idx_56"
                dist_max: 0.5
                dist_min: 0.0


      multi_tof_right_50_58:
        inputs:
        - "multi_tof_right_idx_50"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.10
              low_obstacle:
                input: "multi_tof_right_idx_58"
                dist_max: 0.5
                dist_min: 0.0

      multi_tof_right_52_60:
        inputs:
        - "multi_tof_right_idx_52"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.10
              low_obstacle:
                input: "multi_tof_right_idx_60"
                dist_max: 0.5
                dist_min: 0.0

      multi_tof_right_54_62:
        inputs:
        - "multi_tof_right_idx_54"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.10
              low_obstacle:
                input: "multi_tof_right_idx_62"
                dist_max: 0.5
                dist_min: 0.0
